src/atomicskiplist.ml

```diff
@@ -0,0 +1,221 @@
+type 'a markable_reference = { node : 'a; marked : bool }
+(** markable reference: stores a reference to a node and has a field to specify if it is marked *)
+
+type 'a node = {
+  key : int;
+  height : int;
+  next : 'a node option markable_reference Atomic.t array;
+}
+
+type 'a t = { head : 'a node }
+
+let max_height = 10
+
+(** create_new_node: creates a new node with some value and height *)
+let create_new_node value height =
+  let next = Array.init (max_height + 1) (fun _ -> Atomic.make {node = None; marked = false;})
+  in
+  { key = value; height; next }
+
+(** create_dummy_node_array: Creates a new array with the different node for each index *)
+let create_dummy_node_array () =
+  let arr = Array.init (max_height + 1) (fun _ -> create_new_node Int.max_int max_height)
+  in
+  arr
+
+(** Get a random level from 1 till max_height (both included) *)
+let get_random_level () =
+  let rec count_level cur_level =
+    if cur_level == max_height || Random.float 1.0 <= 0.5 then cur_level
+    else count_level (cur_level + 1)
+  in
+  count_level 1
+
+(** Create a new skiplist *)
+let create () =
+  let tail =
+    {
+      key = Int.max_int;
+      height = max_height;
+      next =
+        Array.make (max_height + 1)
+          (Atomic.make { node = None; marked = false });
+    }
+  in
+  let next = Array.init (max_height + 1) (fun _ -> Atomic.make { node = Some tail; marked=false; }) 
+  in
+  let head = { key = Int.min_int; height = max_height; next } in
+  { head }
+
+(** get_mark_ref: Returns the node and the mark from an Atomic markablereference *)
+let get_mark_ref atomic_ref =
+  let ref = Atomic.get atomic_ref in
+  (ref.node, ref.marked)
+
+(** get_ref: Returns only the node from an Atomic markablereference *)
+let get_ref atomic_ref =
+  let ref = Atomic.get atomic_ref in
+  ref.node
+
+(** Compares old_node and old_mark with the atomic reference and if they are the same then 
+    Replaces the value in the atomic with node and mark *)
+let compare_and_set_mark_ref atomic old_node old_mark node mark =
+  let current = Atomic.get atomic in
+  let set_mark_ref () =
+    Atomic.compare_and_set atomic current { node; marked = mark }
+  in
+  match current.node, old_node with 
+  | Some current_node, Some old_node -> (
+    match node with 
+    | Some node -> current_node == old_node && current.marked = old_mark && ((current_node == node && current.marked = mark) || set_mark_ref ())
+    | None -> false
+  )
+  | None,_|_,None -> false
+
+(** Returns true if key is found within the skiplist else false;
+    Irrespective of return value, fills the preds and succs array with 
+    the predecessors nodes with smaller key and successors nodes with greater than 
+    or equal to key 
+  *)
+let find_in key preds succs sl =
+  let head = sl.head in
+  let init level =
+    let prev = Some head in
+    let curr = get_ref (Option.get prev).next.(level) in
+    let succ, mark = get_mark_ref (Option.get curr).next.(level) in
```

Comment:
This could be just `let {node = succ; marked = mark} = Atomic.get (Option.get curr).next.(level)`.

Also, `Option.get` should generally be avoided.